include/alpaka/core/RuntimeMacros.hpp

```diff
@@ -0,0 +1,38 @@
+/* Copyright 2022  Andrea Bocci, Mehmet Yusufoglu
+ * SPDX-License-Identifier: MPL-2.0
+ */
+
+#pragma once
+
+//! ALPAKA_DEVICE_THROW either aborts(terminating the program and creating a core dump) or throws std::runtime_error
+//! depending on the Acc. It is impossible to "catch" aborts; but the std::runtime_error exception can be catched in
```

Comment:
While we cannot catch the `abort()` or `__trap()` on the device, those should result in raising an asynchronous error in the GPU runtime; we should be able to detect that and throw and exception, that _can_ be caught.

Is this model not working as expected ?

Reply:
I tried to get errors using cudaDeviceSynchronize  and cudaGetLastError but I can not get the error type.  But I have not tried async handlers. Because they will abort the test ?

For example; I used a signal handler for OpemMp to check whether OpenMP converts runtime_error to abort. Yes signal handler catched the SIGABRT signal. But if  I add that handler, the test would abort eventually.

Reply:
I don't have any experience with OpenMP, so I cannot comment in that regard.
However, using signal handlers seems way outside the scope of this feature and of this test.

Reply:
For the CUDA case, here is an example that illustrates the behaviour I have in mind:
```c++
#include <iostream>
#include <string>

#include <alpaka/alpaka.hpp>

using Dim = alpaka::DimInt<1u>;
using Idx = uint32_t;
using Vec = alpaka::Vec<Dim, Idx>;

struct KernelWithThrow {
  template<typename TAcc>
  ALPAKA_FN_ACC void operator()(TAcc const& acc) const {
    auto idx = alpaka::getIdx<alpaka::Grid, alpaka::Threads>(acc);
    if (idx == Vec::zeros()) {
#if defined(ALPAKA_ACC_GPU_CUDA_ENABLED) && defined(__CUDA_ARCH__)
      __trap();
#elif defined(ALPAKA_ACC_GPU_HIP_ENABLED) && defined(__HIP_DEVICE_COMPILE__)
      abort();
#elif defined(ALPAKA_ACC_SYCL_ENABLED) && defined(__SYCL_DEVICE_ONLY__)
      abort();
#else
      throw std::runtime_error();
#endif
    }
  }
};


int main(void) {
  using Platform = alpaka::PlatformCudaRt;
  using Device = alpaka::DevCudaRt;
  using Queue = alpaka::Queue<Device, alpaka::Blocking>;
  using Tag = alpaka::TagGpuCudaRt;

  Platform platform{};
  Device device = alpaka::getDevByIdx(platform, 0);
  Queue queue{device};

  auto const workDiv = alpaka::WorkDivMembers<Dim, Idx>{Vec{8}, Vec{1}, Vec{1}};

  try {
    alpaka::exec<alpaka::TagToAcc<Tag, Dim, Idx>>(queue, workDiv, KernelWithThrow{});
  } catch(std::runtime_error& e) {
    std::cerr << "Caught an error in alpaka::exec:\n" << e.what() << '\n';
  }

  // note that this kind of error is "sticky", and will not be reset by a call to cudaGetLastError()

  try {
    alpaka::wait(queue);
  } catch(std::runtime_error& e) {
    std::cerr << "Caught an error in alpaka::wait:\n" << e.what() << '\n';
  }

  try {
    alpaka::wait(queue);
  } catch(std::runtime_error& e) {
    std::cerr << "Caught an error in alpaka::wait:\n" << e.what() << '\n';
  }

  return 0;
}
```

Compiled with
```bash
/usr/local/cuda/bin/nvcc -std=c++17 -O3 -g -I/home/fwyzard/src/alpaka-group/alpaka/include --expt-relaxed-constexpr -arch sm_86 -DALPAKA_ACC_GPU_CUDA_ENABLED test.cu -o test
```

Results in
```bash
./test 
Caught an error in alpaka::wait:
/home/fwyzard/src/alpaka-group/alpaka/include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp(175) 'TApi::streamSynchronize(queue.getNativeHandle())' A previous API call (not this one) set the error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
Caught an error in alpaka::wait:
/home/fwyzard/src/alpaka-group/alpaka/include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp(175) 'TApi::streamSynchronize(queue.getNativeHandle())' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
/home/fwyzard/src/alpaka-group/alpaka/include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp(74) 'TApi::streamSynchronize(m_UniformCudaHipQueue)' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
/home/fwyzard/src/alpaka-group/alpaka/include/alpaka/queue/cuda_hip/QueueUniformCudaHipRt.hpp(75) 'TApi::streamDestroy(m_UniformCudaHipQueue)' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
```

Reply:
I'm not sure why we don't get the error directly in `alpaka::exec` even with a blocking queue - but we do get the error in `alpaka::wait`.

Reply:
So code did not abort in this case?

Reply:
ok I have added wait for Cuda and added a comment to the test. Thanks.